### PR TITLE
format native queries

### LINKED ISSUE #9142: Proposal: Format SQL Query option in the Native Query Editor
## Proposal

One minor annoyance today in using the Native Query Editor in Metabase is the lack of an option to format/prettify queries.

So I propose to add a `Format Query` option to the Native Query Editor(See screenshot below for a mockup). If this is something that is of interest to the community, I can help contribute to the same.

screenshot 2018-12-17 at 4 43 19 pm

:arrow_down: Please click the :+1: reaction instead of leaving a `+1` or `update?` comment

### PR DESCRIPTION
**Pull request by alxnddr**
Closes https://github.com/metabase/metabase/issues/9142

### Description

It is one of my projects for our internal hackathon. This PR adds format SQL query button.
Since it is a big package which is needed only for a single button, I put it in a separate chunk and import dynamically.
The button is hidden for nosql data sources such as Mongo, Druid, Google Analytics, and some sql databases such as sqlite and mssql for which optional clauses break formatting. For other SQL databases if the formatter library supports the specific dialect of given database it will use the dialcet otherwise it will fall back to ANSI SQL formatter.

### How to verify

Ensure formatting works for SQL databases with the [right dialect](https://github.com/metabase/metabase/pull/38113/files#diff-2432af7b436143db65654d7156e601262c9253931a525630770c53d2dddbec2dR67).
Ensure there is no formatting button for nosql databases.
Ensure the formatter library loads only when user clicks on the Format Query button.

### Demo

https://github.com/metabase/metabase/assets/14301985/d81ff9d6-23de-46c6-a951-7a5d33717a22

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

### CODE REVIEW

File: frontend/src/metabase/query_builder/components/NativeQueryEditor/NativeQueryEditor.tsx [hunk 1]
```diff
@@ -723,6 +724,17 @@ export class NativeQueryEditor extends Component<
     );
   };
 
+  formatQuery = async () => {
+    const engine = this.props.question?.database?.()?.engine;
+    const queryText = this._editor?.getValue();
```
Comment (nemanjaglumac) [inline frontend/src/metabase/query_builder/components/NativeQueryEditor/NativeQueryEditor.tsx:729]:
```suggestion
    const query = question.query();
    const engine = Lib.engine(query);
    const queryText = Lib.rawNativeQuery(query);
```

We ~~can~~ should? use MLv2 going forward. Just make sure you guard against `question` potentially being undefined, and all this should work just fine :)

File: frontend/src/metabase/query_builder/components/NativeQueryEditor/utils.ts [hunk 2]
```diff
@@ -54,3 +56,54 @@ export function calcInitialEditorHeight({
   const lines = getVisibleLinesCount({ query, viewHeight });
   return getEditorLineHeight(lines);
 }
+
+const formatSql = async (sql: string, options: FormatOptionsWithLanguage) => {
+  const sqlFormatter = await import(
+    /* webpackChunkName: "sql-formatter" */ "sql-formatter"
```
Comment (kamilmielnik) [inline frontend/src/metabase/query_builder/components/NativeQueryEditor/utils.ts:62]: Looks like a leftover

File: frontend/src/metabase/query_builder/components/NativeQueryEditor/utils.unit.spec.ts [hunk 3]
```diff
@@ -0,0 +1,53 @@
+import {
```
Comment (kamilmielnik) [inline frontend/src/metabase/query_builder/components/NativeQueryEditor/utils.unit.spec.ts:1]: > ### How to verify
> 
> Ensure formatting works for SQL databases with the [right dialect](https://github.com/metabase/metabase/pull/38113/files#diff-2432af7b436143db65654d7156e601262c9253931a525630770c53d2dddbec2dR67).
> Ensure there is no formatting button for nosql databases.
> Ensure the formatter library loads only when user clicks on the Format Query button.

These sound like very good test cases. Let's cover them!

File: frontend/src/metabase/query_builder/components/NativeQueryEditor/utils.ts [hunk 4]
```diff
@@ -54,3 +56,54 @@ export function calcInitialEditorHeight({
...
+    throw new Error(`No formatter dialect for engine ${engine}`);
+  }
+
+  return formatSql(queryText, {
+    language: dialect,
+    tabWidth: 2,
+    keywordCase: "upper",
+    linesBetweenQueries: 2,
+    paramTypes: {
+      // Snippets, parameters, nested questions
```
Comment (kamilmielnik) [inline frontend/src/metabase/query_builder/components/NativeQueryEditor/utils.ts:105]: Formatting does not work for queries with [optional clauses](https://www.metabase.com/docs/latest/questions/native-editor/sql-parameters).

**Review by uladzimirdev — APPROVED** [review]
LGTM overall :shipit:

File: frontend/src/metabase/query_builder/components/NativeQueryEditor/utils.ts (hunk 4 above)
Comment (alxnddr) [reply frontend/src/metabase/query_builder/components/NativeQueryEditor/utils.ts]: Good catch! Fixed for engines excluding mssql and sqlite due to the lack of support for it

File: frontend/src/metabase/query_builder/components/NativeQueryEditor/utils.ts [hunk 5]
```diff
@@ -54,3 +56,54 @@ export function calcInitialEditorHeight({
   const lines = getVisibleLinesCount({ query, viewHeight });
   return getEditorLineHeight(lines);
 }
+
+const formatSql = async (sql: string, options: FormatOptionsWithLanguage) => {
+  const sqlFormatter = await import(
+    /* webpackChunkName: "sql-formatter" */ "sql-formatter"
+  );
+  return sqlFormatter.format(sql, options);
```
[the remark this answers is not in the document]
Comment (alxnddr) [reply frontend/src/metabase/query_builder/components/NativeQueryEditor/utils.ts:62]: Yup, that is too much for script that is used in a very specific use case. Our bundle is already gigantic

**Review by kamilmielnik — APPROVED** [review]
Awesome!

File: e2e/test/scenarios/native/native.cy.spec.js [hunk 6]
```diff
@@ -459,6 +459,49 @@ describe("no native access", { tags: ["@external", "@quarantine"] }, () => {
   });
 });
 
+describe("scenarios > question > native", { tags: "@mongo" }, () => {
+  const MONGO_DB_NAME = "QA Mongo4";
+
+  before(() => {
```
Comment (nemanjaglumac) [inline e2e/test/scenarios/native/native.cy.spec.js:465]: Let's either use `beforeEach` or let's lose the `describe block` and put everything inside a single `it()` block.

File: e2e/test/scenarios/native/native.cy.spec.js [hunk 7]
```diff
@@ -459,6 +459,49 @@ describe("no native access", { tags: ["@external", "@quarantine"] }, () => {
   });
 });
 
+describe("scenarios > question > native", { tags: "@mongo" }, () => {
```
Comment (nemanjaglumac) [inline e2e/test/scenarios/native/native.cy.spec.js:462]: Is this section related specifically to the code formatting?

File: frontend/src/metabase/query_builder/components/NativeQueryEditor/NativeQueryEditor.tsx [hunk 8]
```diff
@@ -723,6 +725,16 @@ export class NativeQueryEditor extends Component<
     );
   };
 
+  formatQuery = async () => {
+    const { question } = this.props;
+    const query = question.query();
+    const engine = Lib.engine(query);
+    const queryText = Lib.rawNativeQuery(query);
```
Comment (nemanjaglumac) [inline frontend/src/metabase/query_builder/components/NativeQueryEditor/NativeQueryEditor.tsx:732]: <3

**Review by nemanjaglumac — APPROVED** [review]
Left some inline (nit) comments but overall this is 🔥

**Comment by paoliniluis** [comment]
oh god 49 is gonna be so good

**Comment by a-escontrela** [comment]
Is it going to be available for BigQuery? 
Is it possible to customize it by providing our own rules?

**Comment by alxnddr** [comment]
@a-escontrela yes, it will be available for BigQuery as well. Some custom rules may be added later on but are excluded from this version

**Comment by adeturckheim** [comment]
@alxnddr great feature! Big fan of it except for the default setup of a indent of 2 spaces (VS 4 spaces).
I guess it is a matter of personal tastes, but 4 is easier to read.

Only plan on making indent size configurable soon?

File: frontend/src/metabase/query_builder/components/NativeQueryEditor/utils.ts [hunk 9]
```diff
@@ -54,3 +56,58 @@ export function calcInitialEditorHeight({
...
+
+export function formatQuery(queryText: string, engine: string) {
+  const dialect = getFormatterDialect(engine);
+  if (!dialect) {
+    throw new Error(`No formatter dialect for engine ${engine}`);
+  }
+
+  return formatSql(queryText, {
+    language: dialect,
+    tabWidth: 2,
```
Comment (adeturckheim) [inline frontend/src/metabase/query_builder/components/NativeQueryEditor/utils.ts:102]: Would love to be able to edit that one
